
syntax = "proto3";

package descarteslabs.workflows;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";
import "descarteslabs/common/proto/types/types.proto";
import "descarteslabs/common/proto/typespec/typespec.proto";
import "descarteslabs/common/proto/widgets/widgets.proto";
import "descarteslabs/common/proto/logging/logging.proto";

service XYZAPI {
  rpc CreateXYZ(CreateXYZRequest) returns (XYZ) {}
  rpc GetXYZ(GetXYZRequest) returns (XYZ) {}
  rpc ListXYZ(ListXYZRequest) returns (stream XYZ) {}
  rpc DeleteXYZ(DeleteXYZRequest) returns (google.protobuf.Empty) {}
  rpc GetXYZSessionLogs(GetXYZSessionLogsRequest)
      returns (stream XYZLogRecord) {}
}

message XYZ {
  string id = 1;
  // milliseconds since the UNIX epoch (UTC)
  int64 created_timestamp = 2;
  // milliseconds since the UNIX epoch (UTC)
  int64 updated_timestamp = 3;
  string name = 7;
  string description = 8;
  ResultType type = 9;
  string channel = 10;
  string client_version = 11;

  string serialized_graft = 21;
  Typespec typespec = 25;

  // `parameters` are metadata about the Function's arguments.
  // If `typespec` describes a Function, there must be one `parameters` entry
  // for each positional argument, and its `name` must match the name of the corresponding
  // parameter in the `serialized_graft`.
  // Otherwise, `parameters` must be empty.
  repeated Parameter parameters = 27;

  string user = 23;
  string org = 24;
  // Template string for XYZ tiles like
  // "https://workflows.descarteslabs.com/v0-5/xyz/1234567/{z}/{x}/{y}.png".
  // Compatible with Leaflet, OpenLayers, Python format strings.
  string url_template = 26;
  google.protobuf.Timestamp expires_timestamp = 28;
  bool public = 29;
}

message CreateXYZRequest {
  string name = 1;
  string description = 2;
  string serialized_graft = 3;
  Typespec typespec = 4;
  repeated Parameter parameters = 7;
  string channel = 5;
  string client_version = 6;
  uint32 days_to_expiration = 8;
  bool public = 9;
}

message GetXYZRequest { string xyz_id = 1; }

message ListXYZRequest { bool public = 1; }

message DeleteXYZRequest { string xyz_id = 1; }

message GetXYZSessionLogsRequest {
  string xyz_id = 1;
  string session_id = 2;
  LogRecord.Level level = 3;
  // milliseconds since the UNIX epoch (UTC)
  int64 start_timestamp = 4;
}

message XYZLogRecord {
  LogRecord record = 1;
  string session_id = 2;
}
